File: src/commands/handoff.ts
/**
 * CLI Command: Handoff
 *
 * Generates a non-mutating Markdown or JSON handoff report for the current
 * coordinated Arashi workspace.
 */

import {
  checkAllRepos,
  checkRepoStatus,
  collectStatusWarnings,
  shouldIncludeWorkspaceRootInRepositoryChecks,
  summarizeStatuses,
} from "./status.ts";
import {
  createJsonErrorEnvelope,
  createJsonSuccessEnvelope,
  unknownErrorToJsonError,
  writeJsonEnvelope,
} from "../lib/json-output.ts";
import { loadWorkspaceRepositories } from "../lib/config.ts";
import { exec as gitExec } from "../lib/git.ts";
import { findConfiguredWorkspaceRoots, resolveWorkspaceContext } from "../lib/workspace-context.ts";
import { standaloneWorktrees } from "../lib/standalone.ts";
import { info, error as logError, warn } from "../lib/logger.ts";
import { basename, join, relative, resolve } from "path";
import { realpath } from "fs/promises";
import { Command, Option } from "commander";

type JsonWarning = NonNullable<Parameters<typeof createJsonSuccessEnvelope>[2]>[number];
type RepoStatus = Awaited<ReturnType<typeof checkAllRepos>>[number];

const ZERO = 0;
const USAGE_EXIT_CODE = 2;
const ERROR_EXIT_CODE = 1;

export interface HandoffOptions {
  json?: boolean;
  link?: string[];
  nextCommand?: string[];
  risk?: string[];
  todo?: string[];
  validation?: string[];
  markdown?: boolean;
}

interface HandoffContext {
  links: string[];
  nextCommands: string[];
  risks: string[];
  todos: string[];
  validations: string[];
}

interface HandoffRepositorySummary {
  branch: RepoStatus["branch"];
  changeCount: number;
  defaultBranch: RepoStatus["defaultBranch"];
  error: string | null;
  files: RepoStatus["files"];
  name: string;
  path: string;
  refreshWarning: RepoStatus["refreshWarning"];
  state: "clean" | "dirty" | "error";
}

interface HandoffData {
  callerWorktree?: string;
  context: HandoffContext;
  currentRepository: { name: string; path: string } | null;
  effectiveOptions: {
    format: "json" | "markdown";
  };
  generatedNextCommands: string[];
  mode: "configured" | "standalone";
  repositoryPath?: string;
  repositories: HandoffRepositorySummary[];
  summary: ReturnType<typeof summarizeStatuses> & {
    touchedCount: number;
  };
  workspace: {
    branch: string;
    path: string;
  };
  workspaceRoot?: string;
  worktreesBase?: string;
}

interface BuildHandoffDataInput {
  cwd: string;
  options: HandoffOptions;
  statuses: RepoStatus[];
  workspaceBranch: string;
  workspaceRoot: string;
}

const normalizeArray = (value: string[] | undefined): string[] => value ?? [];

const isInsidePath = (candidate: string, parent: string): boolean => {
  const rel = relative(parent, candidate);
  return rel === "" || (!rel.startsWith("..") && !resolve(rel).startsWith(".."));
};

const detectCurrentRepository = (
  cwd: string,
  statuses: RepoStatus[],
): { name: string; path: string } | null => {
  const current = statuses.reduce<RepoStatus | null>((best, status) => {
    if (!isInsidePath(cwd, status.path)) {
      return best;
    }
    if (!best || status.path.length > best.path.length) {
      return status;
    }
    return best;
  }, null);
  if (!current) {
    return null;
  }

  return { name: current.name, path: current.path };
};

const formatBranch = (status: RepoStatus): string => {
  if (status.branch.isDetached) {
    return "detached HEAD";
  }

  let value = status.branch.localBranch || "unknown";
  if (status.refreshWarning?.kind === "missing-remote-ref") {
    return `${value} (${status.refreshWarning.message})`;
  }
  if (status.branch.remoteBranch) {
    value += ` → ${status.branch.remoteBranch}`;
  }

  const drift: string[] = [];
  if (status.branch.ahead > ZERO) {
    drift.push(`ahead ${status.branch.ahead}`);
  }
  if (status.branch.behind > ZERO) {
    drift.push(`behind ${status.branch.behind}`);
  }
  if (drift.length > ZERO) {
    value += ` (${drift.join(", ")})`;
  }

  return value;
};

const statusState = (status: RepoStatus): HandoffRepositorySummary["state"] => {
  if (status.error) {
    return "error";
  }
  if (status.files.length > ZERO) {
    return "dirty";
  }
  return "clean";
};

const summarizeRepo = (status: RepoStatus): HandoffRepositorySummary => ({
  branch: status.branch,
  changeCount: status.files.length,
  defaultBranch: status.defaultBranch,
  error: status.error,
  files: status.files,
  name: status.name,
  path: status.path,
  refreshWarning: status.refreshWarning,
  state: statusState(status),
});

const collectGeneratedNextCommands = (statuses: RepoStatus[]): string[] => {
  const commands = ["arashi status"];
  if (
    statuses.some(
      (status) =>
        status.error ||
        status.files.length > ZERO ||
        status.branch.ahead > ZERO ||
        status.branch.behind > ZERO ||
        status.refreshWarning,
    )
  ) {
    commands.push("arashi status --verbose");
  }
  return commands;
};

const resolveConfiguredWorkspaceBranch = async (
  statuses: RepoStatus[],
  configurationRoot: string,
): Promise<string> => {
  const mainRepository = statuses.find((status) => status.name === "Main Repository");
  if (mainRepository) {
    return mainRepository.branch.localBranch || "unknown";
  }

  try {
    const result = await gitExec(["symbolic-ref", "--short", "HEAD"], configurationRoot);
    const branch = result.stdout.trim();
    if (!branch) {
      return "unknown";
    }
    await gitExec(["show-ref", "--verify", `refs/heads/${branch}`], configurationRoot);
    return branch;
  } catch {
    return "unknown";
  }
};

const buildHandoffData = ({
  cwd,
  options,
  statuses,
  workspaceBranch,
  workspaceRoot,
}: BuildHandoffDataInput): HandoffData => {
  const summary = summarizeStatuses(statuses);
  const touchedCount = statuses.filter(
    (status) => status.files.length > ZERO || status.error,
  ).length;

  return {
    context: {
      links: normalizeArray(options.link),
      nextCommands: normalizeArray(options.nextCommand),
      risks: normalizeArray(options.risk),
      todos: normalizeArray(options.todo),
      validations: normalizeArray(options.validation),
    },
    currentRepository: detectCurrentRepository(cwd, statuses),
    effectiveOptions: {
      format: options.json ? "json" : "markdown",
    },
    generatedNextCommands: collectGeneratedNextCommands(statuses),
    mode: "configured",
    repositories: statuses.map((status) => summarizeRepo(status)),
    summary: {
      ...summary,
      touchedCount,
    },
    workspace: {
      branch: workspaceBranch,
      path: workspaceRoot,
    },
    workspaceRoot,
  };
};

const markdownList = (items: string[], emptyText: string): string => {
  if (items.length === ZERO) {
    return `- ${emptyText}\n`;
  }
  return `${items.map((item) => `- ${item}`).join("\n")}\n`;
};

const markdownChecklist = (items: string[], emptyText: string): string => {
  if (items.length === ZERO) {
    return `- [ ] ${emptyText}\n`;
  }
  return `${items.map((item) => `- [ ] ${item}`).join("\n")}\n`;
};

const formatRepositoryLine = (repo: HandoffRepositorySummary): string => {
  const stateLabel = repo.state === "clean" ? "clean" : repo.state;
  const parts = [`${repo.name}: ${stateLabel}`, `branch ${formatBranch(repo as RepoStatus)}`];
  if (repo.changeCount > ZERO) {
    parts.push(`${repo.changeCount} changed file${repo.changeCount === 1 ? "" : "s"}`);
  }
  if (repo.error) {
    parts.push(repo.error);
  }
  if (repo.refreshWarning && repo.refreshWarning.kind !== "missing-remote-ref") {
    parts.push(repo.refreshWarning.message);
  }
  if (repo.defaultBranch?.state === "available" && repo.defaultBranch.behind > ZERO) {
    parts.push(`default ${repo.defaultBranch.branch} behind by ${repo.defaultBranch.behind}`);
  }
  if (repo.defaultBranch?.state === "unavailable") {
    parts.push(`default ${repo.defaultBranch.branch} unavailable`);
  }
  return `- ${parts.join("; ")}`;
};

export const renderMarkdownReport = (data: HandoffData): string => {
  const touchedRepos = data.repositories.filter((repo) => repo.state !== "clean");
  const repoLines = data.repositories.map(formatRepositoryLine).join("\n");
  const touchedLines =
    touchedRepos.length > ZERO
      ? `${touchedRepos.map((repo) => formatRepositoryLine(repo)).join("\n")}\n`
      : "- All managed repositories are clean.\n";
  const userNextCommands = data.context.nextCommands;
  const generatedNextCommands = data.generatedNextCommands.filter(
    (command) => !userNextCommands.includes(command),
  );
  const commandBlock = [...userNextCommands, ...generatedNextCommands]
    .map((command) => `\`${command}\``)
    .join("\n");

  return `# Arashi Handoff Report

## Workspace

- Workspace mode: ${data.mode}
- Path: ${data.workspace.path}
- Branch: ${data.workspace.branch}
- Caller worktree: ${data.callerWorktree ?? "not applicable"}
- Current repository: ${data.currentRepository ? `${data.currentRepository.name} (${data.currentRepository.path})` : "not resolved"}

## Summary

- Repositories: ${data.summary.total} total, ${data.summary.cleanCount} clean, ${data.summary.dirtyCount} dirty/error
- Touched repositories: ${data.summary.touchedCount}

## Repository Status

${repoLines}

## Repositories Needing Attention

${touchedLines}
## Related Links

${markdownList(data.context.links, "No related links supplied.")}
## Validation Evidence

${markdownList(data.context.validations, "No validation evidence supplied. Add commands/results before relying on this report for merge readiness.")}
## Remaining Work

${markdownChecklist(data.context.todos, "No remaining work supplied.")}
## Risks / Blockers

${markdownList(data.context.risks, "No risks or blockers supplied.")}
## Suggested Next Commands

${commandBlock ? `${commandBlock}\n` : "- No next commands suggested.\n"}
`;
};

const runHandoff = async (options: HandoffOptions): Promise<void> => {
  if (!options.json && options.markdown) {
    warn("--markdown is deprecated; omit --markdown and use the default Markdown output.");
  }
  let context;
  try {
    context = await resolveWorkspaceContext();
  } catch (error) {
    if (options.json)
      writeJsonEnvelope(createJsonErrorEnvelope("handoff", unknownErrorToJsonError(error)));
    else logError(error instanceof Error ? error.message : String(error));
    process.exit(ERROR_EXIT_CODE);
  }
  if (context.mode === "standalone") {
    try {
      const worktrees = await standaloneWorktrees(context);
      const statuses = await Promise.all(
        worktrees.map((worktree) =>
          checkRepoStatus(worktree.branch ?? basename(worktree.path), worktree.path),
        ),
      );
      const callerWorktree = process.cwd();
      const canonicalCaller = await realpath(callerWorktree);
      const canonicalStatusPaths = await Promise.all(
        statuses.map(async (status) => await realpath(status.path)),
      );
      const callerStatus = statuses[canonicalStatusPaths.indexOf(canonicalCaller)];
      const selectedCallerStatus =
        callerStatus ??
        statuses.find((status) => resolve(status.path) === resolve(callerWorktree)) ??
        statuses[ZERO];
      const data: HandoffData = {
        ...buildHandoffData({
          cwd: canonicalCaller,
          options,
          statuses,
          workspaceBranch: selectedCallerStatus?.branch.localBranch || "unknown",
          workspaceRoot: context.mainRoot,
        }),
        callerWorktree,
        currentRepository: selectedCallerStatus
          ? { name: selectedCallerStatus.name, path: selectedCallerStatus.path }
          : null,
        mode: "standalone",
        repositoryPath: context.mainRoot,
        workspace: {
          branch: selectedCallerStatus?.branch.localBranch || "unknown",
          path: context.mainRoot,
        },
        worktreesBase: join(context.mainRoot, ".worktrees"),
      };
      if (options.json)
        writeJsonEnvelope(
          createJsonSuccessEnvelope(
            "handoff",
            data as unknown as Record<string, unknown>,
            collectStatusWarnings(statuses),
          ),
        );
      else process.stdout.write(renderMarkdownReport(data));
      if (statuses.some((status) => status.error !== null)) process.exit(ERROR_EXIT_CODE);
      return;
    } catch (error) {
      if (options.json)
        writeJsonEnvelope(createJsonErrorEnvelope("handoff", unknownErrorToJsonError(error)));
      else logError(error instanceof Error ? error.message : String(error));
      process.exit(ERROR_EXIT_CODE);
    }
  }
  let workspaceRoots;
  try {
    workspaceRoots = await findConfiguredWorkspaceRoots("handoff");
  } catch {
    const message = "Not in an arashi workspace";
    if (options.json) {
      writeJsonEnvelope(
        createJsonErrorEnvelope("handoff", {
          code: "NOT_IN_WORKSPACE",
          message,
        }),
      );
    } else {
      logError(message);
      info("Run 'arashi init' to initialize a workspace before creating a handoff report");
    }
    process.exit(USAGE_EXIT_CODE);
  }

  try {
    const { config } = await loadWorkspaceRepositories(workspaceRoots);
    const includeWorkspaceRoot = await shouldIncludeWorkspaceRootInRepositoryChecks(
      workspaceRoots.executionRoot,
    );
    const statuses = await checkAllRepos(
      workspaceRoots.executionRoot,
      config,
      false,
      includeWorkspaceRoot,
    );
    const workspaceBranch = await resolveConfiguredWorkspaceBranch(
      statuses,
      workspaceRoots.configurationRoot,
    );
    const data = buildHandoffData({
      cwd: process.cwd(),
      options,
      statuses,
      workspaceBranch,
      workspaceRoot: workspaceRoots.configurationRoot,
    });
    data.worktreesBase = resolve(
      workspaceRoots.configurationRoot,
      config.worktreesDir ?? "../.worktrees",
    );
    const warnings: JsonWarning[] = collectStatusWarnings(statuses);

    if (options.json) {
      writeJsonEnvelope(
        createJsonSuccessEnvelope<Record<string, unknown>>(
          "handoff",
          data as unknown as Record<string, unknown>,
          warnings,
        ),
      );
    } else {
      process.stdout.write(renderMarkdownReport(data));
    }

    if (statuses.some((status) => status.error !== null)) {
      process.exit(ERROR_EXIT_CODE);
    }
  } catch (error) {
    if (options.json) {
      writeJsonEnvelope(createJsonErrorEnvelope("handoff", unknownErrorToJsonError(error)));
    } else {
      logError(error instanceof Error ? error.message : String(error));
    }
    process.exit(ERROR_EXIT_CODE);
  }
};

const collectRepeated = (value: string, previous: string[] = []): string[] => [...previous, value];

export const createCommand = (): Command => {
  const deprecatedMarkdown = new Option(
    "--markdown",
    "Deprecated compatibility spelling for the default Markdown output",
  ).hideHelp();
  (deprecatedMarkdown as Option & { deprecated?: boolean }).deprecated = true;

  return new Command("handoff")
    .description("Generate a Markdown or JSON handoff report for the current workspace")
    .option("-j, --json", "Output a structured JSON envelope instead of Markdown")
    .addOption(deprecatedMarkdown)
    .option(
      "--link <link>",
      "Related issue, PR, spec, or reference link (repeatable)",
      collectRepeated,
    )
    .option(
      "--validation <entry>",
      "Validation command and result evidence, e.g. 'pnpm run test — passed' (repeatable)",
      collectRepeated,
    )
    .option(
      "--todo <item>",
      "Remaining work item to include as a checklist entry (repeatable)",
      collectRepeated,
    )
    .option(
      "--risk <item>",
      "Known risk or blocker to include in the report (repeatable)",
      collectRepeated,
    )
    .option(
      "--next-command <command>",
      "Suggested next command, not executed (repeatable)",
      collectRepeated,
    )
    .addHelpText(
      "after",
      `
Examples:
  $ arashi handoff --link https://github.com/corwinm/arashi-arashi/issues/186
  $ arashi handoff --validation "pnpm run test — passed" --todo "watch CI"
  $ arashi handoff --json --next-command "arashi status --verbose"
      `,
    )
    .action(runHandoff);
};

File: src/rust/handoff.rs
//! Non-mutating handoff reports built from the native status snapshot.
use crate::{Result, cli::Args, config::Workspace};
use serde::ser::{Serialize, SerializeMap, SerializeSeq, Serializer};
use serde_json::{Value, json};
use std::path::Path;

pub fn handoff(workspace: &Workspace, cwd: &Path, args: &Args) -> Result<Value> {
    let status = crate::status::status(workspace, cwd)?;
    Ok(build(&status, cwd, args))
}

fn build(status: &Value, cwd: &Path, args: &Args) -> Value {
    let standalone = status["mode"] == "standalone";
    let rows = status[if standalone {
        "worktrees"
    } else {
        "repositories"
    }]
    .as_array()
    .cloned()
    .unwrap_or_default();
    let repositories: Vec<Value> = rows.iter().map(summarize_repository).collect();
    let touched = repositories
        .iter()
        .filter(|row| row["state"] != "clean")
        .count();
    let selected = if standalone {
        let caller = crate::paths::canonicalize(cwd).ok();
        rows.iter()
            .find(|row| {
                caller.as_ref().is_some_and(|caller| {
                    row["path"]
                        .as_str()
                        .is_some_and(|path| Path::new(path) == caller)
                })
            })
            .or_else(|| rows.first())
    } else {
        current_repository(cwd, &rows)
    };
    let workspace_branch = if standalone {
        selected
            .and_then(|row| row["branch"]["localBranch"].as_str())
            .filter(|branch| !branch.is_empty())
            .unwrap_or("unknown")
    } else {
        rows.iter()
            .find(|row| row["name"] == "Main Repository")
            .and_then(|row| row["branch"]["localBranch"].as_str())
            .filter(|branch| !branch.is_empty())
            .unwrap_or("unknown")
    };
    let workspace_root = status["workspaceRoot"].clone();
    let current = selected.map_or(Value::Null, |row| {
        let path = row["path"]
            .as_str()
            .and_then(|path| crate::paths::canonicalize(path).ok())
            .map_or_else(|| row["path"].clone(), |path| json!(path));
        json!({"name":row["name"],"path":path})
    });
    let mut data = json!({
        "context": {
            "links": values(args, "link"),
            "nextCommands": values(args, "next-command"),
            "risks": values(args, "risk"),
            "todos": values(args, "todo"),
            "validations": values(args, "validation")
        },
        "currentRepository": current,
        "effectiveOptions": {"format":if args.has("json") {"json"} else {"markdown"}},
        "generatedNextCommands": generated_commands(&repositories),
        "mode": status["mode"],
        "repositories": repositories,
        "summary": {
            "cleanCount":status["summary"]["cleanCount"],
            "dirtyCount":status["summary"]["dirtyCount"],
            "total":status["summary"]["total"],
            "touchedCount":touched
        },
        "workspace":{"branch":workspace_branch,"path":workspace_root},
        "workspaceRoot":workspace_root,
        "worktreesBase":status["worktreesBase"]
    });
    if standalone {
        data["callerWorktree"] = json!(cwd);
        data["repositoryPath"] = status["repositoryPath"].clone();
    }
    data
}

fn values(args: &Args, key: &str) -> Vec<String> {
    args.options.get(key).cloned().unwrap_or_default()
}

fn summarize_repository(row: &Value) -> Value {
    let mut summary = row.clone();
    if let Some(path) = row["path"].as_str()
        && let Ok(path) = crate::paths::canonicalize(path)
    {
        summary["path"] = json!(path);
    }
    let count = row["files"].as_array().map_or(0, Vec::len);
    summary["changeCount"] = json!(count);
    summary["state"] = json!(if !row["error"].is_null() {
        "error"
    } else if count > 0 {
        "dirty"
    } else {
        "clean"
    });
    summary
}

fn current_repository<'a>(cwd: &Path, rows: &'a [Value]) -> Option<&'a Value> {
    let cwd = crate::paths::canonicalize(cwd).ok()?;
    rows.iter()
        .filter(|row| {
            row["path"]
                .as_str()
                .is_some_and(|path| cwd.starts_with(Path::new(path)))
        })
        .max_by_key(|row| row["path"].as_str().map_or(0, str::len))
}

fn generated_commands(rows: &[Value]) -> Vec<&'static str> {
    let mut commands = vec!["arashi status"];
    if rows.iter().any(needs_verbose_status) {
        commands.push("arashi status --verbose");
    }
    commands
}

fn needs_verbose_status(row: &Value) -> bool {
    !row["error"].is_null()
        || row["files"]
            .as_array()
            .is_some_and(|files| !files.is_empty())
        || row["branch"]["ahead"].as_u64().unwrap_or(0) > 0
        || row["branch"]["behind"].as_u64().unwrap_or(0) > 0
        || comparison_needs_attention(&row["baseBranch"])
        || comparison_needs_attention(&row["defaultBranch"])
        || !row["refreshWarning"].is_null()
}

fn comparison_needs_attention(comparison: &Value) -> bool {
    comparison["state"] == "unavailable"
        || (comparison["state"] == "available" && comparison["behind"].as_u64().unwrap_or(0) > 0)
}

struct OrderedComparison<'a>(&'a Value);

impl Serialize for OrderedComparison<'_> {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        if self.0.is_null() {
            return serializer.serialize_none();
        }
        let mut map = serializer.serialize_map(None)?;
        let keys: &[&str] = match self.0["state"].as_str() {
            Some("available") => &[
                "ahead",
                "behind",
                "branch",
                "compareRef",
                "remote",
                "remoteRef",
                "state",
            ],
            Some("unavailable") if self.0["reason"] == "unresolved-target" => &[
                "branch",
                "compareRef",
                "details",
                "message",
                "reason",
                "remote",
                "remoteRef",
                "state",
            ],
            Some("unavailable") => &[
                "branch",
                "compareRef",
                "remote",
                "remoteRef",
                "details",
                "message",
                "reason",
                "state",
            ],
            _ => &[
                "branch",
                "compareRef",
                "remote",
                "remoteRef",
                "reason",
                "state",
            ],
        };
        for key in keys {
            if let Some(value) = self.0.get(key) {
                map.serialize_entry(key, value)?;
            }
        }
        map.end()
    }
}

struct OrderedRepository<'a>(&'a Value);

impl Serialize for OrderedRepository<'_> {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(None)?;
        map.serialize_entry("baseBranch", &OrderedComparison(&self.0["baseBranch"]))?;
        if let Some(value) = self.0.get("baseBranchSource") {
            map.serialize_entry("baseBranchSource", value)?;
        }
        map.serialize_entry("branch", &self.0["branch"])?;
        map.serialize_entry("changeCount", &self.0["changeCount"])?;
        map.serialize_entry(
            "defaultBranch",
            &OrderedComparison(&self.0["defaultBranch"]),
        )?;
        for key in ["error", "files", "name", "path", "refreshWarning", "state"] {
            map.serialize_entry(key, &self.0[key])?;
        }
        map.end()
    }
}

struct OrderedRepositories<'a>(&'a [Value]);

impl Serialize for OrderedRepositories<'_> {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut sequence = serializer.serialize_seq(Some(self.0.len()))?;
        for repository in self.0 {
            sequence.serialize_element(&OrderedRepository(repository))?;
        }
        sequence.end()
    }
}

struct OrderedHandoffData<'a>(&'a Value);

impl Serialize for OrderedHandoffData<'_> {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(None)?;
        for key in [
            "context",
            "currentRepository",
            "effectiveOptions",
            "generatedNextCommands",
            "mode",
        ] {
            map.serialize_entry(key, &self.0[key])?;
        }
        let repositories = self.0["repositories"]
            .as_array()
            .map_or(&[][..], Vec::as_slice);
        map.serialize_entry("repositories", &OrderedRepositories(repositories))?;
        for key in ["summary", "workspace", "workspaceRoot"] {
            map.serialize_entry(key, &self.0[key])?;
        }
        if self.0["mode"] == "standalone" {
            map.serialize_entry("callerWorktree", &self.0["callerWorktree"])?;
            map.serialize_entry("repositoryPath", &self.0["repositoryPath"])?;
        }
        map.serialize_entry("worktreesBase", &self.0["worktreesBase"])?;
        map.end()
    }
}

struct HandoffEnvelope<'a> {
    data: &'a Value,
    warnings: &'a [Value],
}

impl Serialize for HandoffEnvelope<'_> {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let mut map = serializer.serialize_map(Some(5))?;
        map.serialize_entry("command", "handoff")?;
        map.serialize_entry("data", &OrderedHandoffData(self.data))?;
        map.serialize_entry("ok", &true)?;
        map.serialize_entry("schemaVersion", &1)?;
        map.serialize_entry("warnings", self.warnings)?;
        map.end()
    }
}

pub fn render_json(data: &Value, warnings: &[Value]) -> String {
    serde_json::to_string_pretty(&HandoffEnvelope { data, warnings })
        .expect("handoff values serialize as JSON")
}

pub fn render_markdown(data: &Value) -> String {
    let repositories = data["repositories"].as_array().unwrap();
    let repository_lines = repositories
        .iter()
        .map(repository_line)
        .collect::<Vec<_>>()
        .join("\n");
    let touched = repositories
        .iter()
        .filter(|row| row["state"] != "clean")
        .map(repository_line)
        .collect::<Vec<_>>();
    let touched_lines = if touched.is_empty() {
        "- All managed repositories are clean.\n".to_owned()
    } else {
        format!("{}\n", touched.join("\n"))
    };
    let context = &data["context"];
    let user_commands = strings(&context["nextCommands"]);
    let mut commands = user_commands.clone();
    commands.extend(
        strings(&data["generatedNextCommands"])
            .into_iter()
            .filter(|command| !user_commands.contains(command)),
    );
    let command_block = if commands.is_empty() {
        "- No next commands suggested.\n".to_owned()
    } else {
        format!(
            "{}\n",
            commands
                .iter()
                .map(|command| format!("`{command}`"))
                .collect::<Vec<_>>()
                .join("\n")
        )
    };
    let current = if data["currentRepository"].is_null() {
        "not resolved".to_owned()
    } else {
        format!(
            "{} ({})",
            text(&data["currentRepository"]["name"]),
            text(&data["currentRepository"]["path"])
        )
    };
    let caller = data["callerWorktree"].as_str().unwrap_or("not applicable");

    format!(
        "# Arashi Handoff Report\n\n## Workspace\n\n- Workspace mode: {}\n- Path: {}\n- Branch: {}\n- Caller worktree: {}\n- Current repository: {}\n\n## Summary\n\n- Repositories: {} total, {} clean, {} dirty/error\n- Touched repositories: {}\n\n## Repository Status\n\n{}\n\n## Repositories Needing Attention\n\n{}\n## Related Links\n\n{}\n## Validation Evidence\n\n{}\n## Remaining Work\n\n{}\n## Risks / Blockers\n\n{}\n## Suggested Next Commands\n\n{}\n",
        text(&data["mode"]),
        text(&data["workspace"]["path"]),
        text(&data["workspace"]["branch"]),
        caller,
        current,
        data["summary"]["total"],
        data["summary"]["cleanCount"],
        data["summary"]["dirtyCount"],
        data["summary"]["touchedCount"],
        repository_lines,
        touched_lines,
        markdown_list(&context["links"], "No related links supplied."),
        markdown_list(
            &context["validations"],
            "No validation evidence supplied. Add commands/results before relying on this report for merge readiness."
        ),
        markdown_checklist(&context["todos"], "No remaining work supplied."),
        markdown_list(&context["risks"], "No risks or blockers supplied."),
        command_block
    )
}

fn repository_line(row: &Value) -> String {
    let mut parts = vec![
        format!("{}: {}", text(&row["name"]), text(&row["state"])),
        format!("branch {}", branch_text(row)),
    ];
    let count = row["changeCount"].as_u64().unwrap_or(0);
    if count > 0 {
        parts.push(format!(
            "{count} changed file{}",
            if count == 1 { "" } else { "s" }
        ));
    }
    if let Some(error) = row["error"].as_str() {
        parts.push(error.to_owned());
    }
    if row["refreshWarning"]["kind"] != "missing-remote-ref"
        && let Some(message) = row["refreshWarning"]["message"].as_str()
    {
        parts.push(message.to_owned());
    }
    let base = &row["baseBranch"];
    let default = &row["defaultBranch"];
    let same = base["compareRef"].as_str().is_some()
        && default["compareRef"].as_str().is_some()
        && base["compareRef"] == default["compareRef"];
    append_comparison(
        &mut parts,
        base,
        if same { "base/default" } else { "base" },
        true,
    );
    if !same {
        append_comparison(&mut parts, default, "default", false);
    }
    format!("- {}", parts.join("; "))
}

fn append_comparison(
    parts: &mut Vec<String>,
    comparison: &Value,
    label: &str,
    use_remote_ref: bool,
) {
    let branch = if use_remote_ref {
        comparison["remoteRef"]
            .as_str()
            .or_else(|| comparison["branch"].as_str())
    } else {
        comparison["branch"].as_str()
    }
    .unwrap_or("");
    if comparison["state"] == "available" {
        let behind = comparison["behind"].as_u64().unwrap_or(0);
        if behind > 0 {
            parts.push(format!("{label} {branch} behind by {behind}"));
        }
    } else if comparison["state"] == "unavailable" {
        parts.push(format!("configured {label} {branch} unavailable"));
    }
}

fn branch_text(row: &Value) -> String {
    if row["branch"]["isDetached"] == true {
        return "detached HEAD".to_owned();
    }
    let local = row["branch"]["localBranch"]
        .as_str()
        .filter(|branch| !branch.is_empty())
        .unwrap_or("unknown");
    if row["refreshWarning"]["kind"] == "missing-remote-ref" {
        return format!("{} ({})", local, text(&row["refreshWarning"]["message"]));
    }
    let mut value = local.to_owned();
    if let Some(remote) = row["branch"]["remoteBranch"].as_str() {
        value.push_str(&format!(" → {remote}"));
    }
    let mut drift = Vec::new();
    let ahead = row["branch"]["ahead"].as_u64().unwrap_or(0);
    let behind = row["branch"]["behind"].as_u64().unwrap_or(0);
    if ahead > 0 {
        drift.push(format!("ahead {ahead}"));
    }
    if behind > 0 {
        drift.push(format!("behind {behind}"));
    }
    if !drift.is_empty() {
        value.push_str(&format!(" ({})", drift.join(", ")));
    }
    value
}

fn strings(value: &Value) -> Vec<String> {
    value
        .as_array()
        .into_iter()
        .flatten()
        .filter_map(Value::as_str)
        .map(str::to_owned)
        .collect()
}

fn markdown_list(value: &Value, empty: &str) -> String {
    let values = strings(value);
    if values.is_empty() {
        format!("- {empty}\n")
    } else {
        format!(
            "{}\n",
            values
                .iter()
                .map(|value| format!("- {value}"))
                .collect::<Vec<_>>()
                .join("\n")
        )
    }
}

fn markdown_checklist(value: &Value, empty: &str) -> String {
    let values = strings(value);
    if values.is_empty() {
        format!("- [ ] {empty}\n")
    } else {
        format!(
            "{}\n",
            values
                .iter()
                .map(|value| format!("- [ ] {value}"))
                .collect::<Vec<_>>()
                .join("\n")
        )
    }
}

fn text(value: &Value) -> &str {
    value.as_str().unwrap_or("")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn refresh_failure_comparison_preserves_source_spread_order() {
        let comparison = json!({
            "branch":"main",
            "compareRef":"refs/remotes/origin/main",
            "details":{"error":"fetch failed","kind":"generic"},
            "message":"fetch failed",
            "reason":"refresh-failed",
            "remote":"origin",
            "remoteRef":"origin/main",
            "state":"unavailable"
        });

        assert_eq!(
            serde_json::to_string(&OrderedComparison(&comparison)).unwrap(),
            r#"{"branch":"main","compareRef":"refs/remotes/origin/main","remote":"origin","remoteRef":"origin/main","details":{"error":"fetch failed","kind":"generic"},"message":"fetch failed","reason":"refresh-failed","state":"unavailable"}"#
        );
    }
}
